scripts/validate-dependencies.js

```diff
@@ -0,0 +1,181 @@
+    ),
+  ]),
+);
+
+/**
+ * A map of dependency names to maps of semver expressions to
+ * lists of dependent packages.
+ *
+ * @example
+ *   Map {
+ *     '@babel/core' => Map {
+ *       '^7.12.0' => [
+ *         {name: '@parcel/transformer-js', type: 'dependencies'},
+ *         {name: '@parcel/babel-preset-env', type: 'devDependencies'},
+ *         {name: '@parcel/babel-preset-env', type: 'peerDependencies'}
+ *       ],
+ *       '^7.12.2' => [
+ *         {name: '@parcel/resolver-default', type: 'devDependencies'}
+ *       ]
+ *     }
+ *   }
+ */
+let dependencyMap = new Map();
+for (let [name, meta] of packages.entries()) {
+  if (PACKAGE_EXCLUDES.has(name)) continue;
+  for (let type of DEPENDENCY_TYPES.keys()) {
+    let dependencies = meta[type];
+    if (dependencies) {
+      for (let [dependencyName, semver] of Object.entries(dependencies)) {
+        if (DEPENDENCY_EXCLUDES.has(dependencyName)) continue;
+        let dependentsMap = dependencyMap.get(dependencyName) || new Map();
+        dependencyMap.set(dependencyName, dependentsMap);
+        let dependents = dependentsMap.get(semver) || [];
+        dependentsMap.set(semver, dependents);
+        dependents.push({name, type});
+      }
+    }
+  }
+}
+
+/**
+ * Validation errors are pushed here instead of being thrown.
+ * After all packages have been examined, if there were any errors,
+ * the errors will be printed to stderr and the process will exit
+ * with a non-zero status code.
+ */
+let errors = [];
+
+function reportVersionMismatch(
+  name,
+  expr,
+  dependents,
+  otherExpr,
+  otherDependents,
+) {
+  for (let dependent of dependents) {
+    for (let otherDependent of otherDependents) {
+      errors.push(
+        new DependencyMismatchError(
+          name,
+          expr,
+          dependent,
+          otherExpr,
+          otherDependent,
+        ),
+      );
+    }
+  }
+}
+
+function reportParcelVersionMismatch(name, expr, dependents, version) {
+  for (let dependent of dependents) {
+    errors.push(new PackageMismatchError(name, expr, dependent, version));
+  }
+}
+
+function validateSemverExpressions(name, dependentsMap) {
+  let [expr, ...otherExprs] = [...dependentsMap.keys()];
+  while (otherExprs.length) {
+    let dependents = dependentsMap.get(expr);
+    for (let otherExpr of otherExprs) {
+      if (!semver.intersects(expr, otherExpr)) {
+        reportVersionMismatch(
+          name,
+          expr,
+          dependents,
+          otherExpr,
+          dependentsMap.get(otherExpr),
+        );
+      }
+    }
+    [expr, ...otherExprs] = otherExprs;
+  }
+}
+
+function validateParcelPackageVersions(name, dependentsMap) {
+  let version = packages.get(name).version;
+  for (let expr of dependentsMap.keys()) {
+    if (!semver.satisfies(version, expr)) {
+      reportParcelVersionMismatch(name, expr, dependentsMap.get(expr), version);
+    }
+  }
+}
+
+/** main */
+for (let [name, dependentsMap] of dependencyMap.entries()) {
```

Comment:
nit:

```suggestion
for (let [name, dependentsMap] of dependencyMap) {
```

(Maps are already iterable and yield their entries)